pkg/api/job_runs.go

```diff
@@ -81,3 +86,182 @@ func JobsRunsReportFromDB(dbc *db.DB, filterOpts *filter.FilterOptions, release
+			return apitype.ProwJobRunFailureAnalysis{}, err
+		}
+		if len(ar) == 0 {
+			return apitype.ProwJobRunFailureAnalysis{}, fmt.Errorf("no releases found in db")
+		}
+
+		compareRelease = ar[0].Release
+	}
+
+	return runJobRunAnalysis(jobRun, compareRelease,
+		func(testName, release, suite string, variants []string) (*apitype.Test, error) {
+
+			logger := log.WithFields(log.Fields{
+				"func":     "testResultsFunc",
+				"jobRunID": jobRun.ID,
+			})
+
+			fil := &filter.Filter{
+				Items: []filter.FilterItem{
+					{
+						Field:    "name",
+						Not:      false,
+						Operator: filter.OperatorEquals,
+						Value:    testName,
+					},
+				},
+				LinkOperator: "and",
+			}
+			trs, _, err := BuildTestsResults(dbc, release, "default", false, false,
+				fil)
+			if err != nil {
+				return nil, err
+			}
+			logger.Infof("Got test results: %d", len(trs))
+			for _, tr := range trs {
+				// TODO: this is a weird way to get the variant we want, but it allows re-use
+				// of the existing code.
+				if reflect.DeepEqual(tr.Variants, variants) && tr.SuiteName == suite {
+					return &tr, nil
+				}
+			}
+
+			return nil, nil
+		})
+}
+
+// testResultsFunc is used for injecting db responses in unit tests.
+type testResultsFunc func(testName string, release, suite string, variants []string) (*apitype.Test, error)
+
+func runJobRunAnalysis(jobRun *models.ProwJobRun, compareRelease string,
+	testResultsFunc testResultsFunc) (apitype.ProwJobRunFailureAnalysis, error) {
+
+	logger := log.WithFields(log.Fields{
+		"func":     "jobRunAnalysis",
+		"jobRunID": jobRun.ID,
+	})
+
+	logger.WithField("url", jobRun.URL).Info("loaded prow job run for analysis")
+	logger.Infof("this job run has %d failed tests", len(jobRun.Tests))
+	logger.WithField("variants", jobRun.ProwJob.Variants).Debug("job variants")
+
+	response := apitype.ProwJobRunFailureAnalysis{
+		ProwJobRunID: jobRun.ID,
+		ProwJobName:  jobRun.ProwJob.Name,
+		ProwJobURL:   jobRun.URL,
+		Timestamp:    jobRun.Timestamp,
+		Tests:        []apitype.ProwJobRunTestFailureAnalysis{},
+		OverallRisk: apitype.FailureRisk{
+			Level:   apitype.FailureRiskLevelNone,
+			Reasons: []string{},
+		},
+		OpenBugs: jobRun.ProwJob.Bugs,
+	}
+
+	switch {
+
+	// Return early if no tests failed in this run:
+	case len(jobRun.Tests) == 0:
+		response.OverallRisk.Level = apitype.FailureRiskLevelNone
+		response.OverallRisk.Reasons = append(response.OverallRisk.Reasons,
+			"No test failures found in this job run.")
+		return response, nil
+
+	// Return early if we see mass test failures:
+	case len(jobRun.Tests) > 20:
+		response.OverallRisk.Level = apitype.FailureRiskLevelHigh
+		response.OverallRisk.Reasons = append(response.OverallRisk.Reasons,
+			fmt.Sprintf("%d tests failed in this run: High", len(jobRun.Tests)))
+		return response, nil
+	}
+
+	// Iterate each failed test, query it's pass rates by NURPs, find a matching variant combo, and
+	// see how often we've passed in the last week.
```

Comment:
When we get to the bot scoring failed tests before they are in sippy won't we need the logic within this loop broken out?  Wondering if we should do that now, pass in the testID and Test.Name (bot will have to lookup the ID presumably)

Reply:
Definitely needs to happen, but I'd probably go for doing that refactor once I can find the time to dive into the next card more deeply.